BungeeCord-Patches/0066-Use-Velocity-VarintFrameDecoder.patch

```diff
@@ -0,0 +1,213 @@
+From: xIsm4 <minelatinsoporte@gmail.com>
+Date: Thu, 17 Nov 2022 10:52:42 +0100
+Subject: [PATCH] Use Velocity VarintFrameDecoder
+
+
+diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java
+index c0d37142..bdeab0dc 100644
+--- a/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java
++++ b/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java
+@@ -1,65 +1,78 @@
++/*
++ * Copyright (C) 2018 Velocity Contributors
++ *
++ * This program is free software: you can redistribute it and/or modify
++ * it under the terms of the GNU General Public License as published by
++ * the Free Software Foundation, either version 3 of the License, or
++ * (at your option) any later version.
++ *
++ * This program is distributed in the hope that it will be useful,
++ * but WITHOUT ANY WARRANTY; without even the implied warranty of
++ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
++ * GNU General Public License for more details.
++ *
++ * You should have received a copy of the GNU General Public License
++ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
++ */
++
+ package net.md_5.bungee.protocol;
+ 
++import net.md_5.bungee.protocol.VarintByteDecoder.DecodeResult;
+ import io.netty.buffer.ByteBuf;
+-import io.netty.buffer.Unpooled;
+ import io.netty.channel.ChannelHandlerContext;
+ import io.netty.handler.codec.ByteToMessageDecoder;
+-import io.netty.handler.codec.CorruptedFrameException;
++
+ import java.util.List;
+ 
+-public class Varint21FrameDecoder extends ByteToMessageDecoder
+-{
++public class Varint21FrameDecoder extends ByteToMessageDecoder {
+ 
+-    private static boolean DIRECT_WARNING;
++    private static final FastDecoderException BAD_LENGTH_CACHED =
++            new FastDecoderException("Bad packet length");
++    private static final FastDecoderException VARINT_BIG_CACHED =
++            new FastDecoderException("VarInt too big");
+ 
+     @Override
+-    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception
+-    {
+-        // If we decode an invalid packet and an exception is thrown (thus triggering a close of the connection),
+-        // the Netty ByteToMessageDecoder will continue to frame more packets and potentially call fireChannelRead()
+-        // on them, likely with more invalid packets. Therefore, check if the connection is no longer active and if so
+-        // sliently discard the packet.
+-        if ( !ctx.channel().isActive() )
+-        {
+-            in.skipBytes( in.readableBytes() );
++    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
++        if (!ctx.channel().isActive()) {
++            in.clear();
+             return;
+         }
+ 
+-        in.markReaderIndex();
++        final VarintByteDecoder reader = new VarintByteDecoder();
```

Comment:
allocating object on every decode, for what?

Reply:
Actually on HotSpot JVM this is optimised

Reply:
I specifically crafted `VarintByteDecoder` in such a fashion so that the HotSpot JVM would optimize away the object creation through inlining the code in `AbstractByteBuf.forEachByte()`.

Appearances can be quite deceiving.

Reply:
> Appearances can be quite deceiving!

I will left it like this then

Reply:
> I specifically crafted `VarintByteDecoder` in such a fashion so that the HotSpot JVM would optimize away the object creation through inlining the code in `AbstractByteBuf.forEachByte()`, allowing escape analysis to do the rest of the work.
> 
> Appearances can be quite deceiving!

Inmy opinion its unneccessary fancyness youre doing there.

Reply:
It's not fanciness, it was me observing the behavior of the JVM and guiding it in the best direction in optimizing the code.

Reply:
So you say the overall shorter and still readable while loop is not getting to the same optimizations?